### DISCUSSION TITLE
Add Support for Naming Tabs and Displaying Tab Name/Number on Switch

### DISCUSSION THREAD

**Discussion by BBennettChiba** [ghostty-org/ghostty#3709] [Feature Requests, Ideas] [answered]
I would like to request the following feature:

Tab Naming: Allow tabs to be given custom names, with default names based on their number (e.g., "1," "2," etc.) if no custom name is provided.
Tab Switch Display: When switching between tabs, briefly display the tab's name (or number, if unnamed) in the center of the screen for half a second. This visual feedback would make it easier to confirm the current tab during fast tab navigation.
This functionality would greatly improve the usability and organization of tabs, especially in workflows involving multiple tabs with distinct purposes.

Is this feature already supported? If not, could it be considered for future development? Thank you for your attention to this request!

**Comment by haphamdev** [comment]
Looking for this as well. `goto_tab` only support numeric index now. It would be really nice to be able to funzzy find the tab by name.

Context: I usually work with different repositories simultaneously, each on on tab, switching back-and-force between repos by name is critical for me.

**Reply by haphamdev** [reply]
Wezterm and Tmux support this really smoothly.

**Reply by rhodes-b** [reply]
you can actually search tabs by name on linux with the tab overview look at this example

https://github.com/user-attachments/assets/372787f4-bff0-415b-a93c-2272f77a65aa

I didn't see someone else commented this below
https://github.com/ghostty-org/ghostty/discussions/3709#discussioncomment-11715031

**Comment by EngineersBox** [comment]
Adding to this, It might be handy to support a dotfile in a directory such as `.tabconf` which can specify the name of the tab programmatically, and determine if it should apply to all child directories if another `.tabconf` doesn't exist in the child path. I Added a similar thing to my kitty install so that I could name each project without needing to manually alter the tab name each time.

This may well be something that should be supported via some kind of extension via the up and coming library for the terminal, but I think it's worth noting as part of this discussion.

**Reply by rhodes-b** [reply]
#2353 will let you programatically control ghostty I think this covers your usecase

**Comment by goproslowyo** [comment]
#### Tab Switch Display
**macOS** has a native tab overview mode `cmd+shift+\` and then you can search the titles just by typing.
On **Linux** libadwaita (iirc) gives us a tab overview toggle. You need to bind `toggle_tab_overview` to any combo you'd like and then you have built-in search based on tab title as well.

No tab naming features here, that is to say. YMMV but hope this helps everyone's use cases a little bit.

**Reply by mattpetters** [reply]
This is so cool, had no idea.

**Comment by zhephree** [comment]
I usually have different environments open in each tab (local, dev, staging, prod) and naming them and/or color-coding the tabs would be a huge help to differentiate between environments

**Reply by ghost** [reply]
I reckon a new discussion or an issue, this is really important!

**Comment by 0xrusowsky** [comment]
fyi there is already some wip on https://github.com/ghostty-org/ghostty/issues/2509

**Reply by tkolleh** [reply]
Is there a way to specify the directory and command with these changes? e.g. `nvim ~/.config/`

**Comment by tonykay** [comment]
Yes I really miss this from iTerm (plus the ability to write a badge say top right hand corner). 

My so so compromise is to run tmux in every tab and then the session title appears: eg `tmux new-session -t showroom-llamastack` gives me 

<img width="960" alt="image" src="https://github.com/user-attachments/assets/a37663fb-56b9-4ea5-afd1-462c25299d9d" />

**Reply by devjona** [reply]
I had an iTerm setup/profile for different Rails apps at my job with tabs like this:
`|rails s| rails c | psql | test | git | k9s | extra |` and they were color-coded. It was nice to be able to have that consistently so I could stay organized in my workflow.

**Comment by rhodes-b** [comment]
Macos already shows the tab index - shows this a couple of days after release https://github.com/ghostty-org/ghostty/discussions/3948 (which requests to be able to turn them off)


You can also name tabs already - this only exists on the main branch (you can even rename the tab from the command pallete on macos im not sure of the command pallete state on linux / what it supports I know there is a PR close to being merged last I checked)
https://github.com/ghostty-org/ghostty/pull/4217 (initial tab renaming support macos)
https://github.com/ghostty-org/ghostty/pull/5905 (initial tab renaming support linux)

the only thing I think does not exist here is displaying only on a switch context (afaik ghostty does not have this concept anyways it seems like you want a different implementation than than the tab overview modes)

Also fwiw this has 2 separate feature requests (as such it should have been 2 separate discussions)
naming tabs - this is done
a different switch context which displays the tab number  (since tab overviews aren't what you are looking for)


I don't know how the tab overview works on macos but like I commented above you can even search by tab name on linux
https://github.com/ghostty-org/ghostty/discussions/3709#discussioncomment-13033371

now a question why does not always showing a tab number not satisfy this case? (besides the argument of "visual clutter" which many people care a bit too much about)

**Comment by devjona** [comment]
As an example:

I had an iTerm setup/profile for different Rails apps at my job with tabs like this:
`|rails s| rails c | psql | test | git | k9s | extra |` and they were color-coded. It was nice to be able to have that consistently so I could stay organized in my workflow.

If this feature doesn't exist, it'd be great if we could add it.

**Comment by stabildev** [comment]
renaming tabs would be very helpful when running multiple instances of the same tool

**Reply by rhodes-b** [reply]
you can already do this on tip

**Reply by 00-kat** [reply]
[`prompt_surface_title`](https://tip.ghostty.org/docs/config/keybind/reference#prompt_surface_title) for completeness.

**Reply by cespare** [reply]
To be clear, that names the *surface* (i.e., split), not the whole tab. If you don't use splits, then they're basically the same. But if you do use splits, then that feature doesn't really cut it.

**Reply by rhodes-b** [reply]
> To be clear, that names the surface (i.e., split), not the whole tab. If you don't use splits, then they're basically the same. But if you do use splits, then that feature doesn't really cut it.

https://github.com/ghostty-org/ghostty/discussions/7581 
is another request for this that is more specific about the prompt_surface_title being applied to a tab and not splits.
(like stated above this is also requesting 2 separate features so it should also be 2 separate requests)

**Comment by tobiaskraus** [comment]
You can change the name of your surface (which is basically your tab when you didn't split, if I'm not mistaken). This feature exists since version 1.2.0 (Sept 2025).

Open your Settings and assign it some keybinding. For example CMD+r:

```
keybind = cmd+r=prompt_surface_title
```

Restart Ghostty and then it looks like this:

<img width="838" height="384" alt="image" src="https://github.com/user-attachments/assets/98dbf4ae-8ede-4508-bbd0-58d8e0b661a9" />

**Reply by mirandaio** [reply]
I usually have multiple splits in a tab (text editor, logs, etc). When selecting "Change Terminal Title" from the command palette, it only changes the title for the currently focused split. Is there a way to change the tile for the whole tab so that I don't have to change the title of every split within the tab?

**Reply by cespare** [reply]
@mirandaio You should upvote https://github.com/ghostty-org/ghostty/discussions/7581 as well. (Ever since split naming was added, there has been a lot of confusion where people asking for tab naming have been told "it has already been implemented!"...)

**Comment by shayne** [comment]
I discovered "Change Title" by right clicking in the terminal window. I had previously tried right clicking on the tab itself, expecting it to be there. It would be great to add it to the context menu for the tab itself. Perhaps, even double clicking the existing title lets you edit it. (macOS)

**Reply by joodaloop** [reply]
Yup, would be perfect.

**Reply by diofeher** [reply]
Agreed, no one is expecting this to be on the terminal window, so on the tab would be better.

**Reply by mitchellh** [reply]
This has been on the tab for awhile on main, fyi. Hiding this as outdated.

**Comment by bernsno** [comment]
This might be a separate feature request, but it'd be valuable for tabs that have titles intentionally set to be able to navigate to them by title rather than by process in the command palette. Quick example attached. (Also just figured out that you can use the command palette to set tab titles now!)

Tab Names:
<img width="576" height="169" alt="image" src="https://github.com/user-attachments/assets/bbdd0dfd-c4a7-47c4-8ba0-44db5148a938" />
 
 Command Palette Results:

<img width="509" height="265" alt="image" src="https://github.com/user-attachments/assets/573a8c35-1442-46e0-9326-d07f46f86861" />

**Reply by bernsno** [reply]
I created a PR for this in case the consensus is that this is valuable: https://github.com/ghostty-org/ghostty/pull/10458

**Reply by mitchellh** [reply]
Merged thanks.

**Comment by TalGreenboard** [comment]
PLEASEEEEE

**Comment by Elijen** [comment]
For what it's worth you can name all the splits the same (Right Click -> Change Title...) and the tab will always show this title. Still would be nice to be able to override the tab name.

**Reply by cgestes** [reply]
doing this... quite annoying... would love to automate it for my 10+ tabs.. (with each having at least 3 splits)

**Reply by mitchellh** [reply]
You can do this now on main. Hiding this as outdated. Right click the tab and change tab title (or menu, or command palette, or binding).

**Comment by mitchellh** [comment, marked as answer]
Tabs can now be renamed. This discussion is polluted with a bunch of other unrelated our outdated comments, so I suggest for other features noted here to open a new discussion. Thank you!

**Reply by nmassi** [reply]
How?

**Reply by mitchellh** [reply]
I'm sorry, if you have a paid support contract please reach out to your account manager.

Otherwise, please use search engines.

**Reply by oitray** [reply]
what a horrible reply. I'll use a different project then.

**Reply by mitchellh** [reply]
Lmao. Bye, felicia.

**Reply by mitchellh** [reply]
Imagine someone spending 3 years giving something away for free and for fun, and the recipient not willing to spend minutes searching and then getting mad about that. Open source, baby!

**Reply by 00-kat** [reply]
> > Tabs can now be renamed.
>
> How?

For the record, you can bind something to [`prompt_surface_title`] (split-specific title) or [`prompt_tab_title`] (tab-specific title). Those are also available in the [command palette]. You can also right-click on a tab or the terminal surface/split to access a button from the context menu. On Linux, the hamburger menu on the headerbar also has it; I'm not sure about its equivalent on macOS, but I believe the menu in the bar (or whatever it's called) has it somewhere too.

<details><summary><em>(Hopefully neutral) responses to other comments minimized to detract attention; click or tap to expand.</em></summary>

> I'll use a different project then.

You are of course free to use whatever you want, provided you find an alternative that suits your needs; Ghostty isn't keen on *forcing* you into using it.

> Otherwise, please use search engines.
> [⋮]
> not willing to spend minutes searching

To be fair nothing about nmassi's comment says they hadn't already searched, and neither Google nor DuckDuckGo get you too far (the closest they got me was the [configuration reference], which is positively enormous and has no hints). The website also has no search option (see #3195). [Going through a discussion search for “rename tab”], hovering over them to see if the clipped content looks relevant, puts the first useful result (#9897) around halfway down the first page for me, with everything else looking completely unrelated (usually I give up on a search if half the first page is nonsense). The only related thing I found above it was #10791, but the preview in the search page made it look irrelevant (I only clicked on it as I had answered it myself). Despite knowing where this is documented, I also wasn't able to find it on the [keybind action reference] page initially either, since I was too focused on the word “rename” while the docs use “change”; I only found it because I had “prompt” memorized. So I don't think it's too unreasonable to ask—I too would have given up on searching by now, had I not already known the answer.

Of course, that's not to say nmassi asking here was perfectly fine: the discussion rules (pinned on the discussions page) state “do not revive old posts”, so nmassi *should* have opened a new discussion about this, which Mitchell already hinted toward in his initial comment. I'm personally empathetic toward those who hadn't noticed them (perhaps they found this while searching Google itself; this page was the first result for me), but it's very understandable for open source maintainers to find it frustrating, especially as they're doing this in their spare time for free.

I also don't intend to say that open source maintainers should be responsible for dealing with an endless barrage of (perhaps poorly worded) support requests, especially when in the wrong forum or in violation of the space's rules; I don't blame anyone here for their opinion on this topic.

</details>

To avoid further argument on this (lest it get too toxic, which I feel it is already teetering dangerously close toward), and as every comment emails multiple dozen people subscribed to this thread (see #6937 § 7), I'm **going to lock** this discussion. **If you have more questions, [open a new discussion]!**

[`prompt_surface_title`]: https://ghostty.org/docs/config/keybind/reference#prompt_surface_title
[`prompt_tab_title`]: https://ghostty.org/docs/config/keybind/reference#prompt_tab_title
[command palette]: https://ghostty.org/docs/install/release-notes/1-2-0#command-palette
[configuration reference]: https://ghostty.org/docs/config/reference
[Going through a discussion search for “rename tab”]: https://github.com/ghostty-org/ghostty/discussions?discussions_q=rename+tab
[keybind action reference]: https://ghostty.org/docs/config/keybind/reference
[open a new discussion]: https://github.com/ghostty-org/ghostty/discussions/new?category=q-a
